### PR TITLE
fix: updated navigation event handling

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR separates navigation handling and lifecycle handling in Puppeteer:
- The new `NavigationWatcher` strictly handles navigation events and has been simplified drastically to only watch for the navigation response which appears after the request (hence the navigation) has finished.
- There is also some refactoring for tidiness.

### CODE REVIEW

File: src/common/Page.ts [hunk 1]
```diff
@@ -1871,7 +1871,9 @@ export class Page extends EventEmitter {
   async reload(options?: WaitForOptions): Promise<HTTPResponse | null> {
     const result = await Promise.all([
       this.waitForNavigation(options),
-      this.#client.send('Page.reload'),
+      this.evaluate(() => {
+        window.location.reload();
```
Comment (OrKoN) [inline src/common/Page.ts]: window.location might be overloaded by the page? Why not keep using Page.reload?

Comment (OrKoN) [reply src/common/Page.ts]: another thing discovered: `window.location.reload` would not work for data URLs.

File: src/common/DOMWorld.ts [hunk 2]
```diff
@@ -284,17 +284,10 @@ export class DOMWorld {
...
-      waitUntil,
-      timeout
+      timeout,
+      waitUntil
     );
-    const error = await Promise.race([
-      watcher.timeoutOrTerminationPromise(),
-      watcher.lifecyclePromise(),
-    ]);
-    watcher.dispose();
```
Comment (OrKoN) [inline src/common/DOMWorld.ts:294]: do we still need dispose?

Comment (jrandolf) [reply src/common/DOMWorld.ts:294]: No, all watchers dispose themselves on error or return.

File: src/common/FrameManager.ts [hunk 3]
```diff
@@ -206,56 +207,16 @@ export class FrameManager extends EventEmitter {
...
-          ? new Error(`${response.errorText} at ${url}`)
-          : null;
-      } catch (error) {
-        if (isErrorLike(error)) {
-          return error;
-        }
-        throw error;
-      }
-    }
+    await this.#navigate(url, referer, frame._id);
```
Comment (OrKoN) [inline src/common/FrameManager.ts:215]: I think we should call waitForFrameNavigation before we call navigate?

File: test/src/navigation.spec.ts [hunk 4]
```diff
@@ -116,11 +116,19 @@ describe('navigation', function () {
         window.addEventListener(
           'beforeunload',
           () => {
-            return history.replaceState(null, 'initial', window.location.href);
+            history.replaceState(null, 'initial', window.location.href);
           },
           false
         );
       });
+      // TODO: There is a determinism problem here. Because the history API is
```
Comment (OrKoN) [inline test/src/navigation.spec.ts:124]: is this comment still relevant?

**Comment by OrKoN** [comment]
Closed due to lack of activity.

**Pull request closed without merging**